### ISSUE TITLE
Complementary filter with magnetometer data gives a highly unstable yaw angle

### ISSUE THREAD

**Issue by SahanGura** [CCNYRoboticsLab/imu_tools#148]
I have been working with this imu tools package for a month now. First I could configure the complementary filter with magnetometer data. I used calibrated data from both magnetometer and accelerometer when using the complementary filter and it perfectly worked. 
But now when I am following the same procedure after reassembling the imu to my device, complimentary filter with magnetometer data gives a very unstable yaw angle. When magnetometer is not used, yaw angle is stable to a certain level. Can someone explain why would this happen? I tried calibrating the magnetometer several times but it does not support.

**Comment by mintar** [comment]
Maybe there are now magnetic fields from electric motors or similar that disturb the magnetometer?

Perhaps you could plot the magnetometer data using PlotJuggler and see if it makes sense.

**Comment by SahanGura** [comment]
I calibrated the magnetometer data after attaching all the components to the device. So that might not be an issue of magnetic interferences right? Without using magnetometer data complimentary filter outputs a good orientation estimation. But when using magnetometer, yaw angle get distorted (As the paper says mag data is only used to adjust the yaw estimation from gyro. Therefore distortion of yaw angle is obvious when magnetometer data is not  good). Might be an issue with my magnetometer of the imu. I should change the device and check again.

If you have any other suggestions that would be really appreciated.

**Comment by mintar** [comment]
You can calibrate out static magnetic fields (like metallic parts of the robot itself), but not dynamic magnetic fields (like electric motors, or metallic parts in the environment). Personally I never use the magnetometer when I'm working in an indoor environment.

On the other hand, you said it was working perfectly before. So if you can find the difference between "then" and "now", you should be able to solve it. I still recommend plotting the values though.

**Comment by SahanGura** [comment]
Thank you for your explanation and recommendation. I will try to figure this out.

**Comment by robertogl** [comment]
Does your setup allow you to perform tests outdoor? That may be the easiest to narrow down the problem.

**Comment by SahanGura** [comment]
![Screenshot 2021-11-02 at 01 09 39](https://user-images.githubusercontent.com/50650704/139735721-4212d723-df95-4d40-820b-f3ed8da47172.png)
Hello @robertogl & @mintar ,
In this plot you can see imu raw data values when turning the x axis of the imu to point upwards. I guess there is no issue with the imu raw data. Any ideas regarding this?
Still the complimentary filter cannot estimate the yaw angle properly.

**Comment by mintar** [comment]
Could you record a rosbag that allows me to reproduce the problem?

**Comment by dantenoguera** [comment]
Hi! I am currently having the same issue,  orientation has a lot of error when fusing magnetometer. I simulated imu and magnetometer in Gazebo and recorded a rosbag.

[imu_filter.tar.gz](https://github.com/ccny-ros-pkg/imu_tools/files/7488398/imu_filter.tar.gz)

**Comment by mintar** [comment]
@SahanGura : It seems like your magnetic values are much too large; the earth magnetic field at 0° latitude, 0° longitude is 3.2 × 10−5 T (31.869 μT). However, this shouldn't cause the problem, since the vector is normalized anyway inside `imu_filter_madgwick`.

@dantenoguera : Thanks for the test data, I can confirm the problem. As you said, it only occurs when fusing magnetometer data. I looked into it, but I couldn't spot the problem. Any hints are welcome.

**Comment by robertogl** [comment]
Hi, 
I do not currently have ROS installed in this computer. Could someone please plot raw data along with orientation data (better if Euler angles)?
Also, does this problem occur only when using the complementary filter or also with the Madgwick filter?

Thanks.

**Comment by mintar** [comment]
The problem is that when using the magnetometer data, the orientation is unstable. When not using magnetometer data, it's dead stable (see videos + plots below).

This occurs both with the complementary and the Madgwick filter.

### orientation when using mag data

https://user-images.githubusercontent.com/320188/140771889-318d2886-8ecd-49c2-a7c3-aaf4dc3e02fa.mp4

### orientation when not using mag data

https://user-images.githubusercontent.com/320188/140771922-0926c61c-288b-4b0d-b39d-6b677004f329.mp4


## Input data

### angular velocity

![input_angular_velocity](https://user-images.githubusercontent.com/320188/140770036-9872306e-6c7f-4cb5-9871-3ab112c3d443.png)

### linear acceleration

![input_linear_acceleration](https://user-images.githubusercontent.com/320188/140770070-6aa3d704-4a1a-45dc-a3a0-28d1222e0904.png)

### magnetic field

![input_magnetic_field](https://user-images.githubusercontent.com/320188/140770107-b60e9ad2-dd23-405b-9271-9b4b03871327.png)
 
## Output

### orientation when using mag data

![output_orientation](https://user-images.githubusercontent.com/320188/140770159-7b9e8ead-8eac-440b-8607-9f78d4d02d83.png)

### orientation when not using mag data

Note the different scaling on the y axis - the orientation stays almost at (0, 0, 0), which is correct.

![output_orientation_no_mag](https://user-images.githubusercontent.com/320188/140771171-afe76545-0421-4bc9-b924-f70f664ba224.png)

**Comment by robertogl** [comment]
Wait, so pitch is also affected by mag field? That is not possible with the complementary filter.
I see a very noisy ang_vel_y, which explains the noisier pitch but that should not increase if magnetic data is used. Could you just compare the two pitch values (with and without mag)?

In the two videos the big vector is the raw gravity vector from acceleration data, right? In that case, its value should be the same regardless of what are you fusing. However, the gravity vector in the first video looks to be jumping much more than in the other video. Is that just because of the different perspective or is there anything else?

**Comment by mintar** [comment]
> Could you just compare the two pitch values (with and without mag)?

Sure:

![pitch](https://user-images.githubusercontent.com/320188/140913773-2206d987-3667-4c49-8e1f-a93626b60915.png)

> In the two videos the big vector is the raw gravity vector from acceleration data, right? In that case, its value should be the same regardless of what are you fusing. 

Right.

> However, the gravity vector in the first video looks to be jumping much more than in the other video. Is that just because of the different perspective or is there anything else?

Yes, it's just the perspective. The input acceleration has a high noise in the `x` dimension and low noise in the `y` dimension, that's why it looks like it was jumping more in the first video.

**Comment by mintar** [comment]
I still believe there's something fishy in the input data, because both the Madgwick and complementary filter exhibit the same behavior, but I can't figure out what it is.

**Comment by robertogl** [comment]
Yes, the data is strange and I wonder how it was simulated in Gazebo, what kind of motion if any at all. That said, this difference in the pitch angle should not happen anyway. I can guarantee that the math in the complementary filter works such that the magnetic field data does not affect the roll and pitch values no matter how bad the magnetic field is. Thus, if this is happening something must have changed either in the rotation of the magnetic field into the 'horizontal plane' (the part that calculates the 'L' vector), or in the quaternion multiplication. Otherwise some strange change in the axis. 

As far as I know nothing has changed in the complementary_filter.cpp file in years, so the original math should be preserved. 

Just to make sure is not a data problem, we could try the filters with different data, even if the mag field is completely wrong, roll and pitch should not change after magnetic correction for the complementary filter output. Although some variations could be appreciated for the Madgwick filter's roll and pitch.   

Also, I noticed that the launch file provided by dantenoguera uses world_frame = ENU. I wonder if that has any effect at all. Perhaps try to comment it out?

**Comment by dantenoguera** [comment]
The simulated IMU was mounted on a car-like robot which remained still during the recording.

**Comment by SahanGura** [comment]
I have calibrated magnetometer data and recorded a [rosbag](https://drive.google.com/file/d/1gsrt2OTMH65x91cn6BgoXDZgZCh31aJX/view?usp=sharing) which contains raw data in following topics.

- /imu/accelerometer  --> calibrated acc data
- /imu/gyroscope  --> calibrated gyro data
- /imu/magnetometer  --> calibrated mag data

- /imu/data_raw  --> sensor_msgs/Imu
- /imu/mag  --> sensor_msgs/MagneticField

Could you please test these data with fusing magnetometer data and otherwise and check the difference of the orientation estimation. I get a drastically different result. Following figures shows the calibration of mag data and plotted data of the magnetometer.
@mintar @robertogl 

![Screenshot from 2021-11-10 00-54-39](https://user-images.githubusercontent.com/50650704/140998134-9035feda-4158-4165-bd7b-6a693adeee43.png)

![Screenshot from 2021-11-10 01-18-47](https://user-images.githubusercontent.com/50650704/140998215-a10849b7-6c9c-404f-be41-12ac5b2aee16.png)

**Comment by SahanGura** [comment]
These are the complimentary filters results for the above [rosbag](https://drive.google.com/file/d/1gsrt2OTMH65x91cn6BgoXDZgZCh31aJX/view?usp=sharing) without fusing magnetometer data and with fusing magnetometer data respectively.

- Look when mag data is not used, roll & pitch are perfect and the yaw angle difference is also okay for slow and small movements. (without fusing mag data we cannot get the absolute heading)
- But when we fuse mag data, even the roll and pitch affects in some cases (with 90 degree roll or pitch turns). absolute heading is not accurate as well.

https://user-images.githubusercontent.com/50650704/141046620-947eaf39-1b50-4927-addc-3fb6fe7c2f4d.mp4

https://user-images.githubusercontent.com/50650704/141046661-b1576af7-b384-4dc5-b9e4-bdaae21d07de.mp4

**Comment by SahanGura** [comment]
This is the Madgwick filter's output for the above rosbag with fused mag data. For this rosbag; Madgwick output is better than the complementary filter's output. But with sharp changes of orientation Madgwick filter's convergence time is much larger.

https://user-images.githubusercontent.com/50650704/141048186-b19b03be-027b-49d8-9c12-96afb97a221e.mp4

**Comment by mintar** [comment]
@dantenoguera wrote:

> The simulated IMU was mounted on a car-like robot which remained still during the recording.

To be pedantic, the IMU is *almost* stationary (it rotates *very* slowly in the yaw angle, from 0 degrees to 0.1 degrees in 60 seconds, as can be seen from the `imu/data` topic in the bag file). That's not the problem here, just wanted to note it in case somebody wonders why `angular_velocity.y` is not 0.

**Comment by mintar** [comment]
@robertogl wrote:

> Wait, so pitch is also affected by mag field? That is not possible with the complementary filter.

I think I solved the mystery!

The topics in the rosbag are not synchronized; there are about twice as many messages on the `Imu` topic than on the `MagneticField` topic:

```
topics:      camera/imu/sample   19055 msgs    : sensor_msgs/Imu          
             mag                  9626 msgs    : sensor_msgs/MagneticField
```

When running the complementary filter with `use_mag = true`, the topics will be synchronized, so the filter only sees 9626 of the `Imu` messages. For some reason, running on this subset of messages leads to the pitch fluctuations.

I've synchronized the messages and created a new rosbag that only contains the `Imu` and `MagneticField` messages with matching time stamps:

[imu_filter_synced.bag.gz](https://github.com/ccny-ros-pkg/imu_tools/files/7511635/imu_filter_synced.bag.gz)

On this rosbag, the pitch fluctuations are the same no matter if `use_mag` was `true` or `false`. They are caused by the high noise in the `linear_acceleration.x` component.

Solutions:
* Setting `gain_acc = 0.001` (instead of the default of 0.01) solved the problem.
* Alternatively, reduce the noise in the `linear_acceleration.x` component.

In conclusion, there's nothing wrong with fusing the magnetometer data in this case. So we've solved @dantenoguera 's problem. On to @SahanGura 's! :)

**Comment by SahanGura** [comment]
According to your explanation for the @dantenoguera 's problem; synchronization might be the reason for my issue as well. Do we need to publish both /imu/data & /imu/mag with same time stamp? @mintar 
It's great if you can check whether my data has any issue with the time stamps using this [rosbag](https://drive.google.com/file/d/1gsrt2OTMH65x91cn6BgoXDZgZCh31aJX/view?usp=sharing) file.

**Comment by dantenoguera** [comment]
@mintar Thanks for the corrections. I am still having error in the yaw angle with both filters. I can create a new issue since I am more interested in the Madgwick filter because I can choose to work on a ENU world frame.

![yaw_vs_yaw](https://user-images.githubusercontent.com/48937707/141117595-23cab48a-8873-4559-982d-be75e1faa3a7.png)

The image shows camera/imu/sample/rpy/z which is my yaw ground truth and imu/data/rpy/z which is the yaw I am getting from the Madgwick filter (in degrees). The error gets bigger when I change heading.

**Comment by mintar** [comment]
@dantenoguera : Yes, please create a new issue for this. Also please consider reducing the simulated noise on the angular velocity + linear acceleration or at least make it uniform. In your data, it's weird that `angular_velocity.y`, `linear_acceleration.x` and `linear_acceleration.z` have high noise levels whereas the other components have low noise levels. I can't think of a real IMU that behaves like this.

**Comment by robertogl** [comment]
@mintar wrote:

> I think I solved the mystery!
> 
> The topics in the rosbag are not synchronized; there are about twice as many messages on the `Imu` topic than on the 

Great catch @mintar! So the pitch was just filtered out because of the sampling when mag was not used. Awesome, thanks.

**Comment by mintar** [comment]
@SahanGura : 

> synchronization might be the reason for my issue as well.

No, synchronization is not the issue here, I've checked.

> Do we need to publish both /imu/data & /imu/mag with same time stamp?

No, they don't. The most important thing is that the time stamps are correct (as good as you can make them); don't make your driver lie about them on purpose. :)

If the two topics come from the same device and you *know* that they have been recorded at the same time, they should be published with the exact same time stamp. If not, they should individually have the most accurate time stamp that reflects when the data was really recorded. The filters in this repo use an [ApproximateTime](http://wiki.ros.org/message_filters/ApproximateTime) synchronizer to sync the topics as well as possible. This can lead to dropped messages, but that's the best one can do in this case. In your rosbag, the time synchronizer dropped 1024/6302 (= 16%) of messages, which is acceptable.

### Solution to your problem

The rosbag you recorded was really useful for debugging the problem! Maybe it was a coincidence, but having a rotation around the magnetic axis was immensely useful (see below).

I've noticed that with the complementary filter, the estimated orientation rotates in the opposite direction when using the magnetometer data:

https://user-images.githubusercontent.com/320188/141125456-5dceb20a-707c-42aa-8ae0-77469e091e69.mp4

This indicates that the magnetometer data is not consistent with the data reported by the gyroscope and accelerometer. Let's have a look at the plot of your rosbag:

![plot_rosbag2_2021-11-10-01-33-17](https://user-images.githubusercontent.com/320188/141125805-cf2b41c1-6f99-43ca-a233-9d3a7b8ada24.png)

You can see that at time 15-17, there's a rotation around the x axis (`angular_velocity.x > 0`), and the magnetometer data doesn't change. This means that the **x axis must point either to magnetic north** (or magnetic south). Looking at the magnetometer data for that time, you can see that `magnetometer.y > 0` while `x == 0` and `z == 0`. This means the magnetometer thinks that the **y axis was pointing to magnetic north**. By visualizing the orientation of the complementary filter without magnetometer input and correlating it to the magnetometer values from the plot, I came up with the following substitutions:

mag axis in rosbag | correct mag axis
-------------------|-----------------
y                  |  x
x                  |  y
-z                 |  z

Using that, I wrote a small script to fix the rosbag:

```python
#!/usr/bin/env python3
import rosbag
import sys
import rospy

inbag_name = sys.argv[1]
outbag_name = inbag_name.replace('.bag', '-fixed.bag')

with rosbag.Bag(outbag_name, 'w', compression=rosbag.Compression.LZ4) as outbag:
    for topic, msg, t in rosbag.Bag(inbag_name).read_messages():
        if topic.lstrip('/') == 'imu/mag':
            (msg.magnetic_field.x, msg.magnetic_field.y, msg.magnetic_field.z) = (msg.magnetic_field.y, msg.magnetic_field.x, -msg.magnetic_field.z) 
        outbag.write(topic, msg, t)
```

The result looks correct. Without magnetometer, there's some heavy yaw drift (left), which is corrected when fusing magnetometer data (right).

https://user-images.githubusercontent.com/320188/141127505-45450769-7c05-47ab-924c-d4f49eb09ee6.mp4


### Comparison with `imu_filter_madgwick`

Interestingly, the Madgwick filter doesn't seem to be affected as much by the incorrect magnetometer data, I don't know why:

#### With incorrect mag data

https://user-images.githubusercontent.com/320188/141127668-380ec964-ad18-47b3-bab5-2405e69f3cf0.mp4

#### With fixed mag data

https://user-images.githubusercontent.com/320188/141127708-36feaa9c-26cd-4f83-b711-51da5a00ae73.mp4

**Comment by SahanGura** [comment]
Thank you very much @mintar. I could also figure this out and solve the issue just an hour ago. 
Basically the issue was when fusing magnetometer data in the complementary filter, yaw estimation of the filter was totally distorted. But without magnetometer data, roll and pitch angles were good and the yaw angle difference was also good for smooth and slow movements.
I calibrated the magnetometer data and tried several times but it didn't help. Then I tried to plot the magnetometer data from the mpu9250 driver and figured out my device doesn't give magnetometer values for the correct axis. So I have to reconfigure the magnetometer raw values to x,y,z axes properly. Then this issue could be solved and complementary filter gives the absolute heading without any issue.
Once again, thank you for the support. :)

https://user-images.githubusercontent.com/50650704/141128365-99c6af87-1150-4152-a54a-3a540f5a6f71.mov

.

**Comment by robertogl** [comment]
So once again the complementary filter is not the problem :). Thanks so much @mintar for debugging the data and the great intuition!

**Comment by mintar** [comment]
Thanks for helping debug as well, @robertogl !

**Comment by dantenoguera** [comment]
Found my error, I missed the magnetic declination parameter on the Madgwick filter. I wonder why it needs to be specified manually...  Thanks everyone.

**Comment by mintar** [comment]
> Found my error, I missed the magnetic declination parameter on the Madgwick filter. I wonder why it needs to be specified manually...

Because the magnetic declination depends on your geographic location, so it's different for everyone and has to be specified manually.
